### DISCUSSION TITLE
reveal_type of simple class is `type | Unknown`

### DISCUSSION THREAD

**Discussion by ClementPinard** [microsoft/pyright#5583] [Q&A] [answered]
Hi,

I am currently trying to check type hints for my code, which uses fiftyone's Dataset : https://github.com/voxel51/fiftyone/blob/develop/fiftyone/core/dataset.py#L228

Here is a simple code that produces intriguing results for me :

```python
import fiftyone as fo
reveal_type(fo.Dataset("hello"))
```

The result is `type | Unknown`. How does pyright comes to this conclusion and not `fiftyone.Dataset`, while I literally just called the class constructor ? I know fiftyone is not type hinted, but what kind of typing would pyright need to know that this object is indeed an instance of `fiftyone.Dataset` ?

Tested with mypy and got an error message for lack of typehints as well : `error: Skipping analyzing "fiftyone": module is installed, but missing library stubs or py.typed marker`

I conducted the same reveal_type test with an in-house class and got the expected result, so I guess the problem is on fiftyone side

thanks for your help !

**Comment by ClementPinard** [comment]
Did some tests with pyright versions.

Up to 1.1.302, the result with `--lib` option is

`test.py:5:13 - information: Type of "result" is "Dataset"`

From 1.1.303 to 1.1.305, the result is

`test.py:5:13 - information: Type of "result" is "Unknown"`

For 1.1.306 the result is the same as 1.1.302 ???

`test.py:5:13 - information: Type of "result" is "Dataset"`

From 1.1.307 to 1.1.318, the result is

`test.py:5:13 - information: Type of "result" is "type | Unknown"`

So now, I feel like it might be a bit more than just a lack of type hints for fiftyone...

Any insight appreciated !

**Comment by cainmagi** [comment, marked as answer]
I believe that your issue is similar to this: #5488, because the metaclass of `Dataset` you are using, i.e., `DatasetSingleton`, overrides its `__call__(...)` method.
https://github.com/voxel51/fiftyone/blob/3c0cd9d9ab596ad18506bfeb02a3328638eb8832/fiftyone/core/singletons.py#L26

**Comment by ClementPinard** [comment]
Hi thanks, that seems to be the problem indeed ! Gotta learn what metaclasses are to solve this then :smile:

**Reply by cainmagi** [reply]
You are welcome! Eric has suggested a hot fixture to me before:

> Make the metaclass `__call__(...)` method moved to a if branch like this:
> ```python
> import typing
>
>
> class MetaClass:
>
>     if not typing.TYPE_CHECKING:
>         def __call__(cls, ...):
>            ...
> ```

This change will makes `__call__` method undetectable by the type checker, but still take effect during the run time.

I have come up with another possible solution. But that solution requires `ParamSpec` which also had a small issue before. It should have been fixed now. I will try whether my solution works or not when the new stable `Pylance` is available. If it works, I will let you know! :-)
